src/workerd/jsg/jsg.c++

```diff
@@ -278,27 +278,57 @@ JsSymbol Lock::symbolAsyncDispose() {
   return IsolateBase::from(v8Isolate).getSymbolAsyncDispose();
 }
 
-void Lock::adjustExternalMemory(ssize_t amount) {
-  v8Isolate->AdjustAmountOfExternalAllocatedMemory(static_cast<int64_t>(amount));
+void ExternalMemoryAdjuster::maybeDeferAdjustment(v8::Isolate* isolate, int64_t amount) {
+  if (isolate == nullptr || v8::Locker::IsLocked(isolate)) {
+    isolate->AdjustAmountOfExternalAllocatedMemory(-amount);
+  } else {
+    // Otherwise, if we don't have the isolate locked, defer the adjustment to the next
+    // time that we do.
+    auto& jsgIsolate = *reinterpret_cast<IsolateBase*>(isolate->GetData(0));
+    jsgIsolate.deferExternalMemoryDecrement(amount);
+  }
 }
 
-namespace {
-struct ExternalMemoryAdjuster {
-  int64_t size;
-  v8::Isolate* isolate;
-  ExternalMemoryAdjuster(v8::Isolate* isolate, size_t size)
-      : size(static_cast<int64_t>(size)), isolate(isolate) {
-    isolate->AdjustAmountOfExternalAllocatedMemory(size);
-  }
+ExternalMemoryAdjuster::ExternalMemoryAdjuster(v8::Isolate* isolate, int64_t amount)
+    : amount(amount), isolate(isolate) {
+  isolate->AdjustAmountOfExternalAllocatedMemory(amount);
+}
+
+ExternalMemoryAdjuster::ExternalMemoryAdjuster(ExternalMemoryAdjuster&& other) :
+    amount(other.amount), isolate(other.isolate) {
+  other.amount = 0;
+  other.isolate = nullptr;
+}
+
+ExternalMemoryAdjuster& ExternalMemoryAdjuster::operator=(ExternalMemoryAdjuster&& other) {
+  amount = other.amount;
+  isolate = other.isolate;
+  other.amount = 0;
+  other.isolate = nullptr;
+  return *this;
+}
 
-  ~ExternalMemoryAdjuster() noexcept(false) {
-    isolate->AdjustAmountOfExternalAllocatedMemory(-size);
+ExternalMemoryAdjuster::~ExternalMemoryAdjuster() {
+  if (amount != 0 && isolate != nullptr) {
```

Comment:
Seems safe to check `amount != 0` only, right? It should be illegal to have a non-zero amount with a null isolate.

```suggestion
  if (amount != 0) {
```